### DISCUSSION TITLE
intended usage with react?

### DISCUSSION THREAD

**Discussion by ghost** [TheSpyder/rescript-webapi#43] [Q&A] [answered]
When this library is complete, what is the intention for use with React? For example, I don't see bindings for converting from reactevent to Dom event. Is that intentional?

**Comment by LeoLeBras** [comment]
currently, we have to use an `Obj.magic` 🦄 to interact with a DOM element from a `ReactEvent`
is there a solution?

```rescript
  let onScroll = event => {
    open Webapi.Dom
    // TODO: https://github.com/tinymce/rescript-webapi/discussions/43
    let element = event->ReactEvent.UI.currentTarget->Obj.magic
    let scrollHeight = element->Element.scrollHeight
    let clientHeight = element->Element.clientHeight

    let scrollOffset = scrollHeight - clientHeight * 2
    if element->Element.scrollTop > scrollOffset->Int.toFloat {
      // ... some logic
    }
  }

**Reply by illusionalsagacity** [reply]
If you look at https://github.com/rescript-lang/rescript-react/blob/master/src/ReactEvent.res, there's lots of bindings that are commented that they should return the proper `Dom` type, but this would be a breaking change. So if that were to happen this would be a non-issue for `rescript-webapi`, as you could do:

```rescript
  let onScroll = event => {
    open Webapi.Dom
    let element = event->ReactEvent.UI.currentTarget->EventTarget.unsafeAsElement
    let scrollHeight = element->Element.scrollHeight
    let clientHeight = element->Element.clientHeight

    let scrollOffset = scrollHeight - clientHeight * 2
    if element->Element.scrollTop > scrollOffset->Int.toFloat {
      // ... some logic
    }
  }
```

**Comment by spocke** [comment]
This is more for pure dom bindings similar to how typescript has built in dom bindings. Any library specific things like converting to/from react types could be added to a library that depends on this instead. Like a `rescript-webapi-react` package or similar.

**Comment by TheSpyder** [comment, marked as answer]
The ReScript team has been very clear that they would prefer everyone not use bindings, or make their own, for everything DOM. The return type of `currentTarget`, and many other methods, is an open object:
https://github.com/rescript-lang/rescript-react/blob/c3017ec4bbce5847c3b2da8d2d536450a3e2fd6d/src/ReactEvent.resi#L46-L47

So no, unfortunately it will never be possible for this library to play nicely with `rescript-react`.

**Reply by tom-sherman** [reply]
> unfortunately it will never be possible for this library to play nicely with `rescript-react`

As I understand it this is best qualified with "with the current bindings that `rescript-react` has to event handlers". It should be possible for this library and rescript-react to both be backed by the same opaque types provided by the ReScript stdlib, as @illusionalsagacity  points out this is quite possible in a new major version of `rescript-react`.

**Reply by TheSpyder** [reply]
Possible? yes. Likely? never. That has been repeatedly made clear to me by the ReScript team.
